go-controller/pkg/ovn/loadbalancer.go

```diff
@@ -82,45 +84,140 @@ func (ovn *Controller) getLoadBalancerVIPS(
 	return raw, nil
 }
 
+// deleteLoadBalancerVIP removes the VIP as well as any reject ACLs associated to the LB
 func (ovn *Controller) deleteLoadBalancerVIP(loadBalancer, vip string) {
 	vipQuotes := fmt.Sprintf("\"%s\"", vip)
-	stdout, stderr, err := util.RunOVNNbctl("--if-exists", "remove",
-		"load_balancer", loadBalancer, "vips", vipQuotes)
+	stdout, stderr, err := util.RunOVNNbctl("--if-exists", "remove", "load_balancer", loadBalancer, "vips", vipQuotes)
 	if err != nil {
 		klog.Errorf("Error in deleting load balancer vip %s for %s"+
 			"stdout: %q, stderr: %q, error: %v",
 			vip, loadBalancer, stdout, stderr, err)
+		// if we hit an error and fail to remove load balancer, we skip removing the rejectACL
+		return
 	}
+	ovn.RemoveServiceEndpoints(loadBalancer, vip)
+	ovn.deleteLoadBalancerRejectACL(loadBalancer, vip)
+	ovn.RemoveServiceLB(loadBalancer, vip)
 }
 
-func (ovn *Controller) createLoadBalancerVIP(lb string, serviceIP string, port int32, ips []string, targetPort int32) error {
-	klog.V(5).Infof("Creating lb with %s, %s, %d, [%v], %d", lb, serviceIP, port, ips, targetPort)
-
-	// With service_ip:port as a VIP, create an entry in 'load_balancer'
-	// key is of the form "IP:port" (with quotes around)
-	key := fmt.Sprintf(`"%s"`, util.JoinHostPortInt32(serviceIP, port))
+func (ovn *Controller) configureLoadBalancer(lb, serviceIP string, port int32, endpoints string) error {
+	target := fmt.Sprintf(`vips:"%s"="%s"`, util.JoinHostPortInt32(serviceIP, port), endpoints)
 
-	if len(ips) == 0 {
-		_, _, err := util.RunOVNNbctl("remove", "load_balancer", lb,
-			"vips", key)
-		return err
+	out, stderr, err := util.RunOVNNbctl("set", "load_balancer", lb,
+		target)
+	if err != nil {
+		klog.Errorf("Error in configuring load balancer: %s "+
+			"stdout: %q, stderr: %q, error: %v", lb, out, stderr, err)
+		// endpoints are comma separated
+		ovn.SetServiceEndpointsToLB(lb, util.JoinHostPortInt32(serviceIP, port), strings.Split(endpoints, ","))
 	}
+	return err
+}
+
+// createLoadBalancerVIP either creates or updates a load balancer VIP
+// Calls to this method assume that if ips are passed that those endpoints actually exist
+// and thus the reject ACL is removed
+func (ovn *Controller) createLoadBalancerVIP(lb, serviceIP string, port int32, ips []string, targetPort int32) error {
+	klog.V(5).Infof("Creating lb with %s, %s, %d, [%v], %d", lb, serviceIP, port, ips, targetPort)
 
 	var commaSeparatedEndpoints string
-	for i, ep := range ips {
-		comma := ","
-		if i == 0 {
-			comma = ""
+
+	if len(ips) > 0 {
+		for i, ep := range ips {
+			comma := ","
+			if i == 0 {
+				comma = ""
+			}
+			commaSeparatedEndpoints += fmt.Sprintf("%s%s", comma, util.JoinHostPortInt32(ep, targetPort))
 		}
-		commaSeparatedEndpoints += fmt.Sprintf("%s%s", comma, util.JoinHostPortInt32(ep, targetPort))
+
+		// ensure the ACL is removed if it exists
+		ovn.deleteLoadBalancerRejectACL(lb, util.JoinHostPortInt32(serviceIP, port))
```

Comment:
One thing I just realized is since we are storing these mappings in ovn-k8s. If ovn-k8s gets restarted, the map of ACLs to which LB will be gone. So if ovn-k8s was restarted while a service had no endpoints (reject ACL was applied), and then ovn-k8s came back up and endpoints were added we might be in trouble. I should probably add some logic to rebuild the mappings on OVN restart. Maybe that could be done in a follow up patch?

I think we still need the mapping done in ovn-k8s, as polling OVN each time to figure out which ACL applies to what is not efficient.

Reply:
@dcbw indicated to me that shared informer will cause all events to be added back when ovn-k8s restarts, so it should rebuild the cache itself. No reason to address this with a manual cache rebuild.